aten/src/ATen/native/cuda/MaxUnpooling.cu

```diff
@@ -0,0 +1,610 @@
+  auto dW = stride[2];
+
+  auto padT = padding[0];
+  auto padH = padding[1];
+  auto padW = padding[2];
+
+  TensorArg output_arg{output, "output", 1}, self_arg{self, "self", 2},
+      indices_arg{indices, "indices", 3};
+  checkAllSameGPU(
+      "MaxUnpooling3d_forward_out_cuda", {output_arg, self_arg, indices_arg});
+
+  int64_t batchSize;
+  int64_t inputSlices;
+  int64_t inputTime;
+  int64_t inputHeight;
+  int64_t inputWidth;
+
+  if (self.ndimension() == 4) {
+    /* 5D */
+    batchSize = 1;
+    inputSlices = self.size(0);
+    inputTime = self.size(1);
+    inputHeight = self.size(2);
+    inputWidth = self.size(3);
+    output.resize_({inputSlices, outputTime, outputHeight, outputWidth});
+  } else {
+    /* resize output */
+    batchSize = self.size(0);
+    inputSlices = self.size(1);
+    inputTime = self.size(2);
+    inputHeight = self.size(3);
+    inputWidth = self.size(4);
+    output.resize_(
+        {batchSize, inputSlices, outputTime, outputHeight, outputWidth});
+  }
+  auto input_contiguous = self.contiguous();
+  auto indices_contiguous = indices.contiguous();
+
+  auto output_contiguous = output.contiguous();
+  output_contiguous.zero_();
+
+  // Collapse batch and feature dimensions if needed
+  auto input_contiguous_reshaped = input_contiguous;
+  auto indices_contiguous_reshaped = indices_contiguous;
+
+  if (input_contiguous.ndimension() == 5) {
+    input_contiguous_reshaped = input_contiguous.reshape(
+        {input_contiguous.size(0) * input_contiguous.size(1),
+         input_contiguous.size(2),
+         input_contiguous.size(3),
+         input_contiguous.size(4)});
+    indices_contiguous_reshaped = indices_contiguous.reshape(
+        {indices_contiguous.size(0) * indices_contiguous.size(1),
+         indices_contiguous.size(2),
+         indices_contiguous.size(3),
+         indices_contiguous.size(4)});
+  }
+
+  int totalZ = inputTime * inputSlices * batchSize;
+  int offsetZ = 0;
+  dim3 block(32, 8);
+
+  while (totalZ > 0) {
+    dim3 grid(
+        ceilDiv(inputWidth, static_cast<int64_t>(block.x)),
+        ceilDiv(inputHeight, static_cast<int64_t>(block.y)),
+        totalZ > 65535 ? 65535 : totalZ);
+    AT_DISPATCH_ALL_TYPES_AND_HALF(
+        self.type(), "MaxUnpooling3d_forward_kernel", ([&] {
+          MaxUnpooling3d_forward_kernel<<<
+              grid,
+              block,
+              0,
+              at::cuda::getCurrentCUDAStream()>>>(
+              input_contiguous_reshaped.packed_accessor<scalar_t, 4>(),
+              indices_contiguous_reshaped.packed_accessor<int64_t, 4>(),
+              output_contiguous.data<scalar_t>(),
+              outputTime,
+              outputHeight,
+              outputWidth,
+              offsetZ);
+        }));
+    AT_CHECK(
+        cudaGetLastError() == cudaSuccess,
+        "RoiPooling3d_forward_kernel failed with error code ",
+        cudaGetLastError());
+    totalZ -= 65535;
+    offsetZ += 65535;
+  }
+  return output;
```

Comment:
is this correct?  Aren't you returning the output instead of the data your wrote to?  (Can you add a test?)

Reply:
I might be mistaken here, but `output` is passed in as a packed_tensor as the third argument in max_unpooling3d_forward_kernel. In max_unpooling3d_forward_kernel, the third argument, also named `output`, is the variable being written to.

I might be missing something herre, so correct me if I'm wrong.